Assignment/PingProcess.cs

```diff
@@ -0,0 +1,244 @@
+    }
+
+    //Bullet 2 & 3
+    /*Implement PingProcess' async public Task<PingResult> RunAsync(string hostNameOrAddress) ❌✔
+    First implement the public void RunAsync_UsingTaskReturn_Success() test method to test PingProcess.RunAsync() using "localhost" without using async/await. ❌✔
+    Also implement the async public Task RunAsync_UsingTpl_Success() test method to test PingProcess.RunAsync() using "localhost" but this time DO using async/await. ❌✔
+    Add support for an optional cancellation token to the PingProcess.RunAsync() signature. ❌✔ Inside the PingProcess.RunAsync() invoke the token's ThrowIfCancellationRequested() 
+    method so an exception is thrown. ❌✔ Test that, when cancelled from the test method, the exception thrown is an AggregateException ❌✔ with a TaskCanceledException inner 
+    exception ❌✔ using the test methods RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrapping ❌✔and 
+    RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrappingTaskCanceledException ❌✔ respectively.
+     */
+
+    //Implement PingProcess' async public Task<PingResult> RunAsync(string hostNameOrAddress)
+    //Add support for an optional cancellation token to the PingProcess.RunAsync() signature
+    async public Task<PingResult> RunAsync(
+        string hostNameOrAddress, CancellationToken cancellationToken = default)
+    {
+        try
+        {
+            //Inside the PingProcess.RunAsync() invoke the token's ThrowIfCancellationRequested() method so an exception is thrown.
+            cancellationToken.ThrowIfCancellationRequested();
+            PingResult result = await Task.Run<PingResult>(() => { return Run(hostNameOrAddress); }, cancellationToken);
+            return result;
+        }
+        //when cancelled from the test method, the exception thrown is an AggregateException ❌✔ with a TaskCanceledException inner exception
+        catch (OperationCanceledException)
+        {
+            TaskCanceledException tce = new();
+            AggregateException Ae = new(tce);
+            throw Ae;
+        }
+    }
+    //Bullet 4
+    /*
+     * Complete/fix AND test async public Task<PingResult> RunAsync(IEnumerable<string> hostNameOrAddresses, CancellationToken cancellationToken = default) 
+     * which executes ping for an array of hostNameOrAddresses (which can all be "localhost") in parallel, adding synchronization if needed. ❌✔ NOTE:
+    The order of the items in the stdOutput is irrelevant and expected to be intermingled.
+    StdOutput must have all the ping output returned (no lines can be missing) even though intermingled. ❌✔
+     */
+    async public Task<PingResult> RunAsync(IEnumerable<string> hostNameOrAddresses, CancellationToken cancellationToken = default)
+    {
+        int code = 0;
+        StringBuilder sB = new StringBuilder();
+        //adding synchronization if needed.
+        //Synchronize, one at a time
+        var semaphore = new SemaphoreSlim(1);
```

Comment:
Generally with SemaphoreSlim I recommend putting a max
```suggestion
        var semaphore = new SemaphoreSlim(1,1);
```